lib/chef/context.rb

```diff
@@ -0,0 +1,104 @@
+# freeze_string_literal: true
+#
+# Copyright:: Copyright (c) Chef Software Inc.
+# License:: Apache License, Version 2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require_relative "licensing_config"
+
+class Chef
+  class Context
+    class << self
+      # The test kitchen will generate a nonce and sign it with a secret key.
+      # All these details are written to a file in the temp directory with the same name as the secret key.
+      # The file looks like:
+      # nonce:<nonce>
+      # timestamp:<timestamp>
+      # signature:<signature>
+      #
+      # The nonce is valid for 60 seconds from the time it was generated.
+      # The secret key will be passed to chef-infra client through env variable.
```

Comment:
if they *key* is in the environment variable, then the *key* is also the name of the file, which doesn't seem secure, or am I missing something?

Also, while I think the worry about spoofability is overblown here, this is - if I understand it all correctly, and it's totally possible I don't - equally spoofable, because you're using a symmetric key which you then hand to chef-infra, which means anyone can generate a key, generate a nonce, sign the file, then pass the key they generated to chef-infra - that's no less spoofable than just setting `IS_KITCHEN=true` in the environment - would this not do that?

```ruby
#!/usr/bin/env ruby

require 'securerandom'
require 'openssl'

secret = SecureRandom.hex(10)
nonce = SecureRandom.hex(5)
timestamp = Time.new.to_i
message = "#{nonce}:#{timestamp}"
sig = OpenSSL::HMAC.hexdigest("SHA256", secret, message)

open("/tmp/#{secret}", 'w') do |fh|
  fh.write("nonce:#{nonce}\ntimestamp:#{timestamp}\nsignature:#{sig}")
end

ENV['TEST_KITCHEN_CONTEXT'] = secret

exec("/usr/bin/chef-client")
```

Reply:
No, because the caller can always just jam the key in the vault. You can't use a symmetric key here.

In fact, given the client is in Ruby, there's no good reason to try to make this really unspoof-able, the legal ramifications of "you can be sued for doing this" are your best defense. The people who really want to work around this will, and the rest just won't. IMO, just using the existing TK env variables in the standard open-source version are sufficient.

However, let's walk through your options anyway...

You could make a public keypair, embed the public key in chef-infra, and then try to obfuscate the private key in test-kitchen, but even that is a loosing game, someone will extract the private key from TK.

To make it less trivial, you'd create a CA system:
* When a customer buys a license, they run a tool that generates a keypair and a CSR, and Progress then signs that CSR with their CA and gives the client the certificate.
* TK signs the nonce/timestamp with it's private key and hands the cert to chef-infra
* Chef-infra has the latest CA cert embedded in it and validates the cert is signed by Progress, and if so, uses that cert to validate the signature on the nonce

This is a fair amount of work and requires, among other things:
* A way to rotate the CA securely (one way to make that easier is to have an offline "grandfather" CA which signs the active CA, so if you have to rotate the active CA you go get the key from offline storage, create a new sub-CA with it, and then take that key back offline), but even still you have to prepare for the worst-case where your grandfather expires/gets compromised/etc., which means potentially forcing people to upgrade their infra-client to get the new key
* Setting up a whole CA infrastructure

This is the "right" way to do it, but, lets consider how one would get around this.

In config.rb, re-define all these classes, and just do:

```ruby
class Chef
  class Context
    class << self
      def test_kitchen_context?
        true
      end
    end
  end
end
```

And that's it, you're done, screw licenses.

At the end of the day, people are agreeing to the EULA, and are legally bound to not bypass your licensing restrictions. If they want to, they'll just use cinc, it's easier than all this and has no legal liability. If they want support, they'll pay for licenses and they won't buy 10 and set IS_KITCHEN=true on 10000, because they're likely to get caught the first time they ask for support.

Trying to make this unspoofable is a loosing game.

I *HIGHLY* recommend you simply just check for the environment variable that kitchen already sets, and if so, switch to the workstation entitlement. Some very small number of people will abuse it, but it'll be negligible.

Reply:
Yeah, I agree that making this unspoofable will be very difficult. The user can even go to the code, comment on a few lines, and make the licensing enforcement completely disappear. 

While discussing this with George and Clinton, they mentioned that we must make it a little complex for the user to spoof. Let me check with the architects and check their suggestions as well.

Reply:
Feel free to send them my way, as well. I know George well.

My perspective is this:

If I take 2 minutes to look into it, and all we're doing is checking the environment variable Kitchen sets, I can do that, but I'm making a choice to violate the license agreement, it's a trivial amount of work.

If I take 2 minutes to look into it, and we're doing a bunch of complicated sig verification, I can just comment out that code (or monkeypatch it from config.rb). Still trivial amount of work - in fact the exact same amount of work for me, and still the only thing stopping me is whether I'm willing to violate the license agreement.